File: Source/AlphaTab/Platform/Svg/CssFontSvgCanvas.cs
using AlphaTab.Rendering.Glyphs;

namespace AlphaTab.Platform.Svg
{
    /// <summary>
    /// This SVG canvas renders the music symbols by adding a CSS class 'at' to all elements. 
    /// </summary>
    internal class CssFontSvgCanvas : SvgCanvas
    {
        public override void FillMusicFontSymbol(
            float x,
            float y,
            float scale,
            MusicFontSymbol symbol,
            bool centerAtPosition = false)
        {
            if (symbol == MusicFontSymbol.None)
            {
                return;
            }

            FillMusicFontSymbolText(x, y, scale, "&#" + (int)symbol + ";", centerAtPosition);
        }

        public override void FillMusicFontSymbols(
            float x,
            float y,
            float scale,
            MusicFontSymbol[] symbols,
            bool centerAtPosition = false)
        {
            var s = "";
            foreach (var symbol in symbols)
            {
                if (symbol != MusicFontSymbol.None)
                {
                    s += "&#" + (int)symbol + ";";
                }
            }

            FillMusicFontSymbolText(x, y, scale, s, centerAtPosition);
        }

        private void FillMusicFontSymbolText(
            float x,
            float y,
            float scale,
            string symbols,
            bool centerAtPosition = false)
        {
            Buffer.Append("<g transform=\"translate(" + ((int)x - BlurCorrection) + " " + ((int)y - BlurCorrection) +
                          ")\" class=\"at\" ><text");
            if (scale != 1)
            {
                Buffer.Append(" style=\"font-size: " + scale * 100 + "%; stroke:none\"");
            }
            else
            {
                Buffer.Append(" style=\"stroke:none\"");
            }

            if (Color.Rgba != Model.Color.BlackRgb)
            {
                Buffer.Append(" fill=\"" + Color.Rgba + "\"");
            }

            if (centerAtPosition)
            {
                Buffer.Append(" text-anchor=\"" + GetSvgTextAlignment(Model.TextAlign.Center) + "\"");
            }

            Buffer.Append(">" + symbols + "</text></g>");
        }
    }
}

File: src/platform/svg/CssFontSvgCanvas.ts
import { TextAlign } from '@src/platform/ICanvas';
import { SvgCanvas } from '@src/platform/svg/SvgCanvas';
import { MusicFontSymbol } from '@src/model/MusicFontSymbol';

/**
 * This SVG canvas renders the music symbols by adding a CSS class 'at' to all elements.
 */
export class CssFontSvgCanvas extends SvgCanvas {
    public fillMusicFontSymbol(
        x: number,
        y: number,
        relativeScale: number,
        symbol: MusicFontSymbol,
        centerAtPosition?: boolean
    ): void {
        if (symbol === MusicFontSymbol.None) {
            return;
        }
        this.fillMusicFontSymbolText(x, y, relativeScale, `&#${symbol};`, centerAtPosition);
    }

    public fillMusicFontSymbols(
        x: number,
        y: number,
        relativeScale: number,
        symbols: MusicFontSymbol[],
        centerAtPosition?: boolean
    ): void {
        let s: string = '';
        for (const symbol of symbols) {
            if (symbol !== MusicFontSymbol.None) {
                s += `&#${symbol};`;
            }
        }
        this.fillMusicFontSymbolText(x, y, relativeScale, s, centerAtPosition);
    }

    private fillMusicFontSymbolText(
        x: number,
        y: number,
        relativeScale: number,
        symbols: string,
        centerAtPosition?: boolean
    ): void {
        x *= this.scale;
        y *= this.scale;

        this.buffer += `<g transform="translate(${x} ${y})" class="at" ><text`;
        const scale = this.scale * relativeScale;
        if (scale !== 1) {
            this.buffer += ` style="font-size: ${scale * 100}%; stroke:none"`;
        } else {
            this.buffer += ' style="stroke:none"';
        }
        if (this.color.rgba !== '#000000') {
            this.buffer += ` fill="${this.color.rgba}"`;
        }
        if (centerAtPosition) {
            this.buffer += ` text-anchor="${this.getSvgTextAlignment(TextAlign.Center)}"`;
        }
        this.buffer += `>${symbols}</text></g>`;
    }
}
